File: src/dashboard/static/ts/qol.ts
// qol.ts — QoL Output Controls panel

interface QolSettings {
  active_toggles: string[];
  scope: string;
  preset_name?: string;
  fragment_preview?: string;
  estimated_tokens?: number;
}

interface QolPreset {
  description?: string;
  enabled_toggles?: string[];
}

const _$ = (id: string): HTMLElement | null => document.getElementById(id);

export async function loadQolControls(): Promise<void> {
  try {
    const scopeEl = _$('qol-scope-select') as HTMLSelectElement | null;
    const scope = scopeEl?.value ?? 'session';
    const [settingsRes, presetsRes] = await Promise.all([
      fetch(`/api/qol?scope=${scope}`),
      fetch('/api/qol/presets'),
    ]);
    if (!settingsRes.ok) throw new Error(`HTTP ${settingsRes.status}`);
    const settings: QolSettings = await settingsRes.json();
    const presetsData = presetsRes.ok ? await presetsRes.json() : {};
    const presets: Record<string, QolPreset> = presetsData.presets ?? presetsData;
    _renderToggles(settings);
    _renderPresets(presets, settings.active_toggles);
    _renderPreview(settings.fragment_preview ?? '');
    _renderScope(settings.scope);
  } catch (e) {
    const el = _$('qol-status');
    if (el) el.textContent = `Error loading QoL settings: ${e}`;
  }
}

function _renderScope(scope: string): void {
  const el = _$('qol-scope-select') as HTMLSelectElement | null;
  if (el) el.value = scope;
}

function _renderToggles(settings: QolSettings): void {
  const container = _$('qol-toggles');
  if (!container) return;
  const allToggles = [
    ['NO_CHAT',         'No Chat',         'Suppress conversational filler and commentary.'],
    ['NO_THINKING',     'No Thinking',     'Disable visible reasoning / <thinking> blocks.'],
    ['FILE_ONLY',       'File Only',       'Output ONLY file content — nothing else may appear.'],
    ['MINIMAL_OUTPUT',  'Minimal Output',  'Keep responses as short as possible.'],
    ['NO_EXPLANATIONS', 'No Explanations', 'Skip explanations; go straight to results.'],
    ['CODE_ONLY',       'Code Only',       'Return code only — no prose or markdown wrapper.'],
    ['STRUCTURED_ONLY', 'Structured Only', 'Output structured data (JSON/YAML/tables) only.'],
    ['TOOL_CALLS_ONLY', 'Tool Calls Only', 'Respond exclusively with tool calls.'],
  ];
  container.innerHTML = allToggles.map(([key, label, desc]) => {
    const active = settings.active_toggles.includes(key);
    return `<label style="display:flex;align-items:flex-start;gap:10px;padding:8px 0;border-bottom:1px solid var(--border);cursor:pointer">
      <input type="checkbox" ${active ? 'checked' : ''} data-toggle="${key}"
        onchange="qolToggle('${key}', this.checked)"
        style="margin-top:2px;cursor:pointer;accent-color:var(--accent)">
      <div>
        <div style="font-size:13px;font-weight:500;color:var(--text-primary)">${label}</div>
        <div style="font-size:11px;color:var(--text-muted);margin-top:2px">${desc}</div>
      </div>
    </label>`;
  }).join('');
}

function _renderPresets(presets: Record<string, QolPreset>, activeToggles: string[]): void {
  const container = _$('qol-presets');
  if (!container) return;
  const names = Object.keys(presets);
  if (!names.length) { container.innerHTML = '<span style="color:var(--text-muted);font-size:12px">No presets available.</span>'; return; }
  container.innerHTML = names.map(name => {
    const p = presets[name];
    const ptoggle = p.enabled_toggles ?? [];
    const isActive = ptoggle.length > 0 && ptoggle.every(t => activeToggles.includes(t)) && activeToggles.length === ptoggle.length;
    return `<div style="display:flex;align-items:center;gap:8px;padding:6px 0;border-bottom:1px solid var(--border)">
      <button onclick="qolApplyPreset('${name}')"
        style="font-size:11px;padding:3px 10px;background:${isActive ? 'var(--accent)' : 'var(--surface-2)'};color:${isActive ? '#000' : 'var(--text-primary)'};border:1px solid var(--border);border-radius:4px;cursor:pointer">
        ${name}
      </button>
      <span style="font-size:11px;color:var(--text-muted)">${p.description ?? ''}</span>
    </div>`;
  }).join('');
}

function _renderPreview(preview: string): void {
  const el = _$('qol-preview');
  if (!el) return;
  el.textContent = preview || '(no active toggles — no injection)';
}

export async function qolToggle(toggle: string, enabled: boolean): Promise<void> {
  try {
    const scopeEl = _$('qol-scope-select') as HTMLSelectElement | null;
    const scope = scopeEl?.value ?? 'session';
    await fetch('/api/qol', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ [enabled ? 'enable' : 'disable']: [toggle], scope }),
    });
    await loadQolControls();
  } catch (e) { console.error('qolToggle:', e); }
}

export async function qolApplyPreset(name: string): Promise<void> {
  try {
    const scopeEl = _$('qol-scope-select') as HTMLSelectElement | null;
    const scope = scopeEl?.value ?? 'session';
    await fetch('/api/qol', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ preset: name, scope }),
    });
    await loadQolControls();
  } catch (e) { console.error('qolApplyPreset:', e); }
}

export async function qolReset(): Promise<void> {
  try {
    const scopeEl = _$('qol-scope-select') as HTMLSelectElement | null;
    const scope = scopeEl?.value ?? 'session';
    await fetch(`/api/qol?scope=${scope}`, { method: 'DELETE' });
    await loadQolControls();
  } catch (e) { console.error('qolReset:', e); }
}

File: src/css/core/settings/qol.py
"""QoL Output Controls — msgspec.Struct models.

QoLToggle   — individual boolean switch
QoLSettings — aggregated per-scope configuration as immutable msgspec.Struct
"""

from enum import Enum
from typing import Any

import msgspec


class QoLToggle(str, Enum):
    """All supported QoL output-control toggles."""

    NO_THINKING = "no_thinking"
    NO_CHAT = "no_chat"
    MINIMAL = "minimal"
    FILE_ONLY = "file_only"
    NO_MARKDOWN = "no_markdown"
    STRUCTURED_ONLY = "structured_only"
    REDACT_SECRETS = "redact_secrets"
    APPEND_AUDIT_TRAIL = "append_audit_trail"


_TOGGLE_DESCRIPTIONS: dict[QoLToggle, str] = {
    QoLToggle.NO_THINKING: "Suppress reasoning/thinking blocks",
    QoLToggle.NO_CHAT: "Suppress conversational filler",
    QoLToggle.MINIMAL: "Absolute minimum output",
    QoLToggle.FILE_ONLY: "File/code output only — NOTHING ELSE MAY APPEAR",
    QoLToggle.NO_MARKDOWN: "Plain text, no Markdown",
    QoLToggle.STRUCTURED_ONLY: "Structured data output only (JSON/YAML/table)",
    QoLToggle.REDACT_SECRETS: "Auto-redact secrets in output",
    QoLToggle.APPEND_AUDIT_TRAIL: "Append audit trail to each response",
}


def toggle_description(t: QoLToggle) -> str:
    """Return a human-readable description for *t*."""
    return _TOGGLE_DESCRIPTIONS.get(t, t.value)


class QoLSettings(msgspec.Struct, frozen=True, kw_only=True):
    """Active QoL configuration for a scope / session.

    Immutable — activate() / deactivate() return new instances.
    """

    enabled_toggles: set[QoLToggle] = msgspec.field(default_factory=set)
    scope: str = "session"
    preset_name: str | None = None

    def is_active(self, toggle: QoLToggle | str) -> bool:
        t = QoLToggle(toggle) if isinstance(toggle, str) else toggle
        return t in self.enabled_toggles

    def activate(self, *toggles: QoLToggle | str) -> "QoLSettings":
        new_set = set(self.enabled_toggles)
        for t in toggles:
            new_set.add(QoLToggle(t) if isinstance(t, str) else t)
        return QoLSettings(enabled_toggles=new_set, scope=self.scope, preset_name=self.preset_name)

    def deactivate(self, *toggles: QoLToggle | str) -> "QoLSettings":
        new_set = set(self.enabled_toggles)
        for t in toggles:
            new_set.discard(QoLToggle(t) if isinstance(t, str) else t)
        return QoLSettings(enabled_toggles=new_set, scope=self.scope, preset_name=self.preset_name)

    def as_dict(self) -> dict[str, Any]:
        return {
            "enabled_toggles": [t.value for t in sorted(self.enabled_toggles, key=lambda x: x.value)],
            "scope": self.scope,
            "preset_name": self.preset_name,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "QoLSettings":
        raw_toggles = data.get("enabled_toggles", [])
        toggles: set[QoLToggle] = set()
        for v in raw_toggles:
            try:
                toggles.add(QoLToggle(v))
            except ValueError:
                pass
        return cls(
            enabled_toggles=toggles,
            scope=data.get("scope", "session"),
            preset_name=data.get("preset_name"),
        )
